### ISSUE TITLE
Missing yggdrasil.spec file

### ISSUE THREAD

**Issue by jirihnidek** [RedHatInsights/yggdrasil#121]
The git repository of upstream (on GitHub) does not contain `yggdrasil.spec` file. Because of that:

1) It is not possible to create testing rpm and install it to the system
2) It is not possible to do testing of building rpm as part of CI testing

**Comment by subpop** [comment]
It had one, but I removed it when I ported the build system to meson. I have not added such a thing back yet because I don't consider the absence of a spec file to be that important. The software can be built and installed from the source directory by running: `meson setup -Dprefix=/usr builddir && meson compile -C builddir && sudo meson install -C builddir`.

I'm hesitant to carry distribution-specific instructions in the upstream development repository for two reasons:

1) It implies an unnecessary distribution-bias. The software should run on any distribution; if it does not, I consider that a bug. Carrying one distribution's build instructions and not another implies one distribution is more supported than another.
2) The instructions to build an RPM inevitably deviate from the downstream instructions, creating two problems: (1) changes to one file must be analyzed to determine whether they should be applied to one or the other, or both; (2) the built artifact is not the same as the one built by downstream distributions because of the differences aforementioned.

That being said, I do have a work-in-progress branch to add a `.copr` directory back and support compilation into an RPM using COPR: https://github.com/RedHatInsights/yggdrasil/tree/add-copr

There is one annoying quirk that I haven't fixed yet. When I build the src.rpm locally, `%dist` gets set correctly with the contents of `%distprefix` as generated by `%forgemeta`. But when COPR performs the SRPM build, it defines `%distprefix` correctly, but then %dist is empty. I can't see anything in the logs that indicate why. When the x86_64.rpm is built, it does get a value set correctly in `%dist`. I feel like this is a problem with the COPR build host when building the src.rpm.

I cannot see why the src.rpm does not get configured with a populated `%dist` value, while the binary build does. It's not a hug blocker, but it does make browsing the [list of builds](https://copr.fedorainfracloud.org/coprs/linkdupont/yggdrasil/builds/) difficult to identify which build is which.

**Comment by jirihnidek** [comment]
No, I really do not want to use VM snapshots to revert changes in few bits. I have this experience from one project and it is something horrible. No, never again.

Red Hat is main sponsor of `yggdrasil` development, right? Most of developers and contributors uses Fedora or RHEL based systems, right? Thus, I believe that we deserve to have one small configuration file with suffix `.spec` in upstream repository. I don't believe it is too much. Other distributions (if any) can for example maintain their own spec files (or patches of upstream .spec file) in their own distribution repositories.

When project becomes big and complex (and `yggdrasil` has this potential), then ability to build rpm locally from `.spec` file  is important to be able to do relatively effective development and testing of changes in `.spec` file.

Correct me if I'm wrong, but I assume that making changes to `.spec` file in downstream is something like this: You have to:
1. Commit changes
2. Create some temporary tag
3. Push temporary tag
4. Some rpm build systems (COPR, packit, etc.) downloads `yggdrasil-*tar.gz` file from github and tries to create RPM and after all of this you can see result of build.

Or do you use some trick with `rpmbuild`?

**Comment by subpop** [comment]
I spent a little time fiddling around with COPR some more and discovered a workaround to force populating the empty `%dist` macro... by defining it manually in the spec file. This is a good example of the kind of deviation that I dislike when keeping distribution build instructions upstream. Inadvertently migrating this kind of workaround downstream, where it is unnecessary, can lead to unforeseen issues later on as changes are made to the downstream build environment.

```spec
# Manually redefine %%dist to work around an issue in COPR where the build root
# that creates the srpm does not define a value for %%dist. This should *NOT* be
# carried in downstream; this is strictly an upstream/COPR/CI workaround.
%global dist %{distprefix}.fc%{fedora}
```

However, it does embed the commit hash into the NVR of both the source and binary RPMs, so it fixes my final obstacle to adding back the COPR support. I'll open a PR to merge this branch into `main`.

> No, I really do not want to use VM snapshots to revert changes in few bits. I have this experience from one project and it is something horrible. No, never again.

I agree. I don't want this project to prefer one way of development over another. It currently doesn't encourage or require the use of any specific development software or workflow, and I'd like to keep it that way.

> Correct me if I'm wrong, but I assume that making changes to .spec file in downstream is something like this: You have to:
>
>    Commit changes
>    Create some temporary tag
>    Push temporary tag
>    Some rpm build systems (COPR, packit, etc.) downloads yggdrasil-*tar.gz file from github and tries to create RPM and after all of this you can see result of build.
>
> Or do you use some trick with rpmbuild?

There's nothing tricky happening downstream to package it. Just the standard [`fedpkg` workflow](https://docs.fedoraproject.org/en-US/package-maintainers/Package_Maintenance_Guide/#common_fedpkg_commands).

**Comment by subpop** [comment]
I think we found a solution to this through https://github.com/RedHatInsights/yggdrasil/commit/25647b0ea40592844f7848f4d9d3673086dab7ab
